docs/contributors/code/coding-guidelines.md

```diff
@@ -136,10 +136,317 @@ export { __unstableDoTerribleAwfulAction } from './api';
+
+If your experimental API is only meant for the Gutenberg Plugin but not for the next WordPress major release, consider limiting the export to the plugin environment. For example, `@wordpress/components` could do that to receive early feedback about a new Component, but avoid bringing that component to WordPress core:
+
+```js
+if ( IS_GUTENBERG_PLUGIN ) {
+	export { __experimentalFunction } from './experiments';
+}
+```
+
+#### Replace experimental selectors with hooks
+
+Sometimes a non-exported React hook suffices as a substitute for introducing a new experimental selectors:
+
+```js
+// Instead of this:
+	// selectors.js:
+	export function __unstableHasActiveBlockOverlayActive( state, parent ) { /* ... */ }
+	export function __unstableIsWithinBlockOverlay( state, clientId ) {
+		let parent = state.blocks.parents[ clientId ];
+		while ( !! parent ) {
+			if ( __unstableHasActiveBlockOverlayActive( state, parent ) ) {
+				return true;
+			}
+			parent = state.blocks.parents[ parent ];
+		}
+		return false;
+	}
+	// MyComponent.js:
+	function MyComponent({ clientId }) {
+		const { __unstableIsWithinBlockOverlay } = useSelect( myStore );
+		const isWithinBlockOverlay = __unstableIsWithinBlockOverlay( clientId );
+		// ...
+	}
+
+// Consider this:
+	// MyComponent.js:
+	function hasActiveBlockOverlayActive ( selectors, parent ) { /* ... */ }
+	function useIsWithinBlockOverlay( clientId ) {
+		return useSelect( ( select ) => {
+			const selectors = select( blockEditorStore ); 
+			let parent = selectors.getBlockRootClientId( clientId );
+			while ( !!parent ) {
+				if ( hasActiveBlockOverlayActive( selectors, parent ) ) {
+					return true;
+				}
+				parent = selectors.getBlockRootClientId( parent );
+			}
+			return false;
+		});
+	}
+	function MyComponent({ clientId }) {
+		const isWithinBlockOverlay = useIsWithinBlockOverlay( clientId );
+		// ...
+	}
+```
+
+#### Dispatch experimental actions in thunks
+
+Turning an existing public action into a [thunk](/docs/how-to-guides/thunks.md)
+enables dispatching private actions inline:
+
+```js
+export function toggleFeature( scope, featureName ) {
+    return function ( { dispatch } ) {
+		dispatch({ type: '__experimental_BEFORE_TOGGLE' })
+		// ...
+    };
+}
+```
+
+#### Use the `lock()` and `unlock()` API from `@wordpress/experiments` to privately export almost anything
+
+Each `@wordpress` package wanting to privately access or expose experimental APIs can
+do so by opting-in to `@wordpress/experiments`:
+
+```js
+// In packages/block-editor/experiments.js:
+import { __dangerousOptInToUnstableAPIsOnlyForCoreModules } from '@wordpress/experiments';
+export const { lock, unlock } =
+	__dangerousOptInToUnstableAPIsOnlyForCoreModules(
+		'I know using unstable features means my plugin or theme will inevitably break on the next WordPress release.',
+		'@wordpress/block-editor' // The package opting in
```

Comment:
It took me a long long time to realise that this is supposed to be the name of the package _calling_ `__dangerousOptInToUnstableAPIsOnlyForCoreModules`, not the name of the package whose APIs you want to access.

I wonder if there's a way to get the package name from an environment variable or something so that it's clearer.

Probably not. Just noting my confusion nonetheless.

Reply:
Good note! I just updated the docs to clarify that.

As for the automation – both arguments could probably be injected by a custom babel plugin. Brownie points – it would obscure the correct usage even further for extenders trying to reverse-engineer this.